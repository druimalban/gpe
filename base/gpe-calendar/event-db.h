/*
 * Copyright (C) 2002 Philip Blundell <philb@gnu.org>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version
 * 2 of the License, or (at your option) any later version.
 */

#ifndef EVENT_DB_H
#define EVENT_DB_H

#include <glib.h>
#include <time.h>

typedef enum
{
  RECUR_NONE,
  RECUR_DAILY,
  RECUR_WEEKLY,
  RECUR_MONTHLY,
  RECUR_YEARLY
} recur_t;

typedef struct event_details_s
{
  char *summary;
  char *description;
} *event_details_t;

typedef struct event_s
{
  unsigned long uid;

  time_t start;
  unsigned long duration;	/* 0 == instantaneous */
  unsigned long alarm;		/* -1 == none */
  
  struct
  {
    recur_t type;

    unsigned int count;
    unsigned int daymask;	/* bit 0 = Mon, bit 1 = Tue, etc */

    time_t end;			/* 0 == perpetual */
  } recur;

  event_details_t details;
  
  gboolean mark;
} *event_t;

extern gboolean event_db_start (void);
extern gboolean event_db_stop (void);

extern gboolean event_db_add (event_t);
extern gboolean event_db_remove (event_t);

extern event_t event_db_new (void);
extern void event_db_destroy (event_t);

extern event_details_t event_db_alloc_details (event_t);
extern event_details_t event_db_get_details (event_t);
extern void event_db_forget_details (event_t);

extern GSList *event_db_list_for_period (time_t start, time_t end);

#endif
