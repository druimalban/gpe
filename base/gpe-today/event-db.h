/*
 * Copyright (C) 2002 Philip Blundell <philb@gnu.org>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version
 * 2 of the License, or (at your option) any later version.
 */

#ifndef EVENT_DB_H
#define EVENT_DB_H

#include <glib.h>
#include <time.h>
#include <stdlib.h>

typedef enum
{
  RECUR_NONE,
  RECUR_DAILY,
  RECUR_WEEKLY,
  RECUR_MONTHLY,
  RECUR_YEARLY
} recur_type_t;

typedef struct event_details_s
{
  guint ref;

  unsigned long sequence;
  time_t modified;

  char *summary;
  char *description;
} *event_details_t;

#define FLAG_UNTIMED   (1 << 0)
#define FLAG_ALARM     (1 << 1)
#define FLAG_TENTATIVE (1 << 2)

struct calendar_time_s
{
  u_int16_t year;
  u_int8_t month, day;
  u_int8_t hour, minute;
};

typedef time_t calendar_time_t;

typedef struct recur_s
{
  recur_type_t type;
  
  unsigned int count;
  unsigned int increment;
  unsigned long daymask;	/* bit 0 = Mon, bit 1 = Tue, etc */
  
  time_t end;			/* 0 == perpetual */
} *recur_t;

typedef struct event_s
{
  unsigned long uid;

  calendar_time_t start;
  unsigned long duration;	/* 0 == instantaneous */
  unsigned long alarm;		/* seconds before event */
  unsigned long flags;

  recur_t recur;
  
  event_details_t details;
  gboolean mark;
} *event_t;

extern gboolean event_db_start (void);
extern gboolean event_db_stop (void);

extern gboolean event_db_add (event_t);
extern gboolean event_db_remove (event_t);

extern gboolean event_db_flush (event_t ev);

extern event_t event_db_new (void);
extern void event_db_destroy (event_t);

extern event_details_t event_db_alloc_details (event_t);
extern event_details_t event_db_get_details (event_t);
extern void event_db_forget_details (event_t);

extern GSList *event_db_list_for_period (time_t start, time_t end);
extern GSList *event_db_list_for_future (time_t start, guint max);
extern GSList *event_db_untimed_list_for_period (time_t start, time_t end, gboolean yes);
extern void event_db_list_destroy (GSList *);

extern recur_t event_db_get_recurrence (event_t ev);

#endif
