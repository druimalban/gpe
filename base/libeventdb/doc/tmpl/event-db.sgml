<!-- ##### SECTION Title ##### -->
event-db

<!-- ##### SECTION Short_Description ##### -->


<!-- ##### SECTION Long_Description ##### -->
<para>

</para>

<!-- ##### SECTION See_Also ##### -->
<para>

</para>

<!-- ##### SECTION Stability_Level ##### -->


<!-- ##### MACRO SECONDS_IN_DAY ##### -->
<para>

</para>



<!-- ##### MACRO MON ##### -->
<para>

</para>



<!-- ##### MACRO TUE ##### -->
<para>

</para>



<!-- ##### MACRO WED ##### -->
<para>

</para>



<!-- ##### MACRO THU ##### -->
<para>

</para>



<!-- ##### MACRO FRI ##### -->
<para>

</para>



<!-- ##### MACRO SAT ##### -->
<para>

</para>



<!-- ##### MACRO SUN ##### -->
<para>

</para>



<!-- ##### ENUM recur_type_t ##### -->
<para>

</para>

@RECUR_NONE: 
@RECUR_DAILY: 
@RECUR_WEEKLY: 
@RECUR_MONTHLY: 
@RECUR_YEARLY: 

<!-- ##### MACRO FLAG_UNTIMED ##### -->
<para>

</para>



<!-- ##### MACRO FLAG_ALARM ##### -->
<para>

</para>



<!-- ##### MACRO FLAG_TENTATIVE ##### -->
<para>

</para>



<!-- ##### MACRO FLAG_CLONE ##### -->
<para>

</para>



<!-- ##### MACRO FLAG_RECUR ##### -->
<para>

</para>



<!-- ##### STRUCT calendar_time_s ##### -->
<para>

</para>

@date: 
@time: 

<!-- ##### TYPEDEF calendar_time_t ##### -->
<para>

</para>


<!-- ##### MACRO EVENT_DB_USE_MEMCHUNK ##### -->
<para>

</para>



<!-- ##### MACRO event_db__alloc_event ##### -->
<para>

</para>



<!-- ##### MACRO event_db__alloc_recur ##### -->
<para>

</para>



<!-- ##### MACRO event_db__free_event ##### -->
<para>

</para>

@_x: 


<!-- ##### MACRO event_db__free_recur ##### -->
<para>

</para>

@_x: 


<!-- ##### FUNCTION event_db_start ##### -->
<para>

</para>

@Returns: 


<!-- ##### FUNCTION event_db_refresh ##### -->
<para>

</para>

@Returns: 


<!-- ##### FUNCTION event_db_stop ##### -->
<para>

</para>

@Returns: 


<!-- ##### FUNCTION event_db_add ##### -->
<para>

</para>

@Param1: 
@Returns: 


<!-- ##### FUNCTION event_db_remove ##### -->
<para>

</para>

@Param1: 
@Returns: 


<!-- ##### FUNCTION event_db_flush ##### -->
<para>

</para>

@ev: 
@Returns: 


<!-- ##### FUNCTION event_db_clone ##### -->
<para>

</para>

@Param1: 
@Returns: 


<!-- ##### FUNCTION event_db_new ##### -->
<para>

</para>

@Returns: 


<!-- ##### FUNCTION event_db_find_by_uid ##### -->
<para>

</para>

@uid: 
@Returns: 


<!-- ##### FUNCTION event_db_destroy ##### -->
<para>

</para>

@Param1: 


<!-- ##### FUNCTION event_db_alloc_details ##### -->
<para>

</para>

@Param1: 
@Returns: 


<!-- ##### FUNCTION event_db_get_details ##### -->
<para>

</para>

@Param1: 
@Returns: 


<!-- ##### FUNCTION event_db_forget_details ##### -->
<para>

</para>

@Param1: 


<!-- ##### FUNCTION event_db_list_for_period ##### -->
<para>

</para>

@start: 
@end: 
@Returns: 


<!-- ##### FUNCTION event_db_list_alarms_for_period ##### -->
<para>

</para>

@start: 
@end: 
@Returns: 


<!-- ##### FUNCTION event_db_list_for_future ##### -->
<para>

</para>

@start: 
@max: 
@Returns: 


<!-- ##### FUNCTION event_db_untimed_list_for_period ##### -->
<para>

</para>

@start: 
@end: 
@yes: 
@Returns: 


<!-- ##### FUNCTION event_db_list_destroy ##### -->
<para>

</para>

@Param1: 


<!-- ##### FUNCTION event_db_get_recurrence ##### -->
<para>

</para>

@ev: 
@Returns: 


