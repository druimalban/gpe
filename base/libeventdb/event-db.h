/* event-db.h: Event Database Interface.
 * Copyright (C) 2002, 2006 Philip Blundell <philb@gnu.org>
 *               2006, Florian Boor <florian@kernelconcepts.de>
 * Copyright (C) 2006 Neal H. Walfield <neal@walfield.org>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version
 * 2 of the License, or (at your option) any later version.
 */

#ifndef EVENT_DB_H
#define EVENT_DB_H

#include <glib-object.h>
#include <glib.h>
#include <time.h>

#include "event-cal.h"
#include "event.h"

#define ERROR_DOMAIN() g_quark_from_static_string ("libeventdb")
/* If an error occurs and the passed GError is NULL, we propagate
   error by emitting a signal on EDB.  This must be used instead of
   g_set_error.  */
#define SIGNAL_ERROR(edb, gerror, args , ...) \
  do \
    { \
      char *__e = g_strdup_printf (args, ##__VA_ARGS__); \
      if (gerror) \
        g_set_error (gerror, ERROR_DOMAIN (), 0, \
                     "%s: %s", __func__, __e); \
      else \
        { \
          char *buffer = g_strdup_printf ("%s: %s", __func__, __e); \
          g_signal_emit \
            (edb, EVENT_DB_GET_CLASS (edb)->error_signal, 0, buffer); \
          g_free (buffer); \
        } \
      g_free (__e); \
    } \
  while (0)
#define SIGNAL_ERROR_GERROR(edb, dest, src) \
  do \
    { \
      if ((dest)) \
        g_propagate_error ((dest), (src)); \
      else \
        { \
          char *buffer = g_strdup_printf ("%s:%s", __func__, (src)->message); \
          g_error_free ((src)); \
          g_signal_emit \
            ((edb), EVENT_DB_GET_CLASS (edb)->error_signal, 0, buffer); \
          g_free (buffer); \
        } \
    } \
  while (0)

/* Enumerate the events in the event database EDB which MAY occur
   between (PERIOD_START and PERIOD_END].  If ALARMS is true, returns
   those events which have an alarm which MAY go off between
   PERIOD_START and PERIOD_END.  Calls CB on each event until CB
   returns a non-zero value.  A reference to EV is allocated and the
   callback function must consume it.  */
typedef void (*events_enumerate_t) (EventDB *edb,
				    time_t period_start, time_t period_end,
				    gboolean alarms,
				    int (*cb) (EventSource *ev),
				    GError **error);
/* Returns the event with the event id EVENTID or NULL if there is no
   such event.  */
typedef int (*eventid_to_uid_t) (EventDB *edb, const char *eventid,
				 GError **error);
/* Set the default calendar to EC.  */
typedef void (*set_default_calendar_t) (EventDB *edb, EventCalendar *ec,
					GError **error);

/* Create a new event on backing store.  EV is uninitialized; save the
   allocated UID in EV->UID.  */
typedef gboolean (*event_new_t) (EventSource *ev, GError **error);
/* Load event EV's basic data (i.e. not necessarily its details) from
   backing store.  EV->UID is valid.  Returns FALSE if no such event
   exists.  */
typedef gboolean (*event_load_t) (EventSource *ev, GError **error);
/* Load the event's details from backing store.  event_load has
   already been called successfully.  */
typedef void (*event_load_details_t) (EventSource *ev, GError **error);
/* Flush the event to backing store.  If an error occurs, FALSE should
   be returned and an error may be reported in *ERR which the client
   must free using g_free.  */
typedef gboolean (*event_flush_t) (EventSource *ev, GError **error);
/* Mark the event EV as removed (i.e. deleted).  */
typedef void (*event_remove_t) (EventSource *ev, GError **error);

/* List the events which have alarms which have not yet been
   acknowledged.  */
typedef GSList *(*list_unacknowledged_alarms_t) (EventDB *edb, GError **error);
/* Mark EV's alarm as having fired but yet not been acknowledged.  */
typedef void (*event_mark_unacknowledged_t) (EventSource *ev, GError **error);
/* Mark EV's alarm as having been acknowledged.  */
typedef void (*event_mark_acknowledged_t) (EventSource *ev, GError **error);
/* Mark all alarms having fired through T as having been
   acknowledged.  */
typedef void (*acknowledge_alarms_through_t) (EventDB *edb, time_t t,
					      GError **error);

/* Create a new calendar on the backing store using EC as the
   template.  Must initialize EC->UID to an identifier unique to the
   backing store.  */
typedef void (*event_calendar_new_t) (EventCalendar *ec, GError **error);
/* Flush the calendar to backing store.  */
typedef void (*event_calendar_flush_t) (EventCalendar *ec, GError **error);
/* Delete the calendar from the backing store.  The generic framework
   has already taken care that no dangling references are left
   around.  */
typedef void (*event_calendar_delete_t) (EventCalendar *ec, GError **error);
/* List the events in the calendar EC.  If MODIFIED_AFTER is non-zero,
   limit to those events which have been modified on or after
   MODIFIED_AFTER.  If MODIFIED_BEFORE is non-zero, limit to those
   events which have been modified on or before MODIFIED_BEFORE.  */
typedef GSList *(*event_calendar_list_events_t) (EventCalendar *ec,
						 time_t modified_after,
						 time_t modified_before,
						 GError **error);
/* List the deleted events in the calendar EC.  NB: Deleted events
   have EV->DEAD set to true.  */
typedef GSList *(*event_calendar_list_deleted_t) (EventCalendar *ec,
						  GError **error);
/* Flush any deleted events from EC.  */
typedef void (*event_calendar_flush_deleted_t) (EventCalendar *ec,
						GError **error);

typedef struct
{
  GObjectClass gobject_class;

  events_enumerate_t events_enumerate;
  eventid_to_uid_t eventid_to_uid;
  set_default_calendar_t set_default_calendar;

  event_new_t event_new;
  event_load_t event_load;
  event_load_details_t event_load_details;
  event_flush_t event_flush;
  event_remove_t event_remove;

  list_unacknowledged_alarms_t list_unacknowledged_alarms;
  event_mark_unacknowledged_t event_mark_unacknowledged;
  event_mark_acknowledged_t event_mark_acknowledged;
  acknowledge_alarms_through_t acknowledge_alarms_through;

  event_calendar_new_t event_calendar_new;
  event_calendar_flush_t event_calendar_flush;
  event_calendar_delete_t event_calendar_delete;
  event_calendar_list_events_t event_calendar_list_events;
  event_calendar_list_deleted_t event_calendar_list_deleted;
  event_calendar_flush_deleted_t event_calendar_flush_deleted;

  /* Signals.  */
  guint error_signal;
  EventDBError error;

  guint calendar_new_signal;
  EventCalendarNew calendar_new;
  guint calendar_deleted_signal;
  EventCalendarDeleted calendar_deleted;
  guint calendar_reparented_signal;
  EventCalendarReparented calendar_reparented;
  guint calendar_changed_signal;
  EventCalendarChanged calendar_changed;
  guint calendar_modified_signal;
  EventCalendarModified calendar_modified;
  
  guint event_new_signal;
  EventNew event_new_func;
  guint event_removed_signal;
  EventRemoved event_removed;
  guint event_modified_signal;
  EventModified event_modified;

  guint alarm_fired_signal;
  EventDBAlarmFiredFunc alarm_fired;
} EventDBClass;

struct _EventDB
{
  GObject object;

  GHashTable *events;
  guint default_calendar;
  GSList *calendars;

  /* A list of events that need to be flushed to backing store.  */
  GSList *laundry_list;
  /* The idle source.  */
  guint laundry_buzzer;

  /* A list of events which have no user references.  */
  GSList *cache_list;
  /* Number of seconds to leave unreferenced items in memory before
     deallocating them.  */
#ifndef CACHE_EXPIRE
#define CACHE_EXPIRE 120
#endif
  /* A timer source to periodically flush the cache.  */
  guint cache_buzzer;

  /* The list of events with upcoming alarms.  We hold a reference to
     each.  */
  GSList *upcoming_alarms;
  /* EVENTS contains alarms until this point in time.  */
  time_t period_end;

  /* The alarm source.  */
  guint alarm;

  /* The point through which alarms have been fired (when an alarm
     fires, it is entered into the alarms_unacknowledged table and
     only removed once it has been acknowledged).  */
  time_t alarms_fired_through;
};

/* E is an Event or an EventCalendar and is dirty (i.e. needs to be
   written to disk).  Add it to the laundry list and do it when we are
   idle.  */
extern void add_to_laundry_pile (GObject *e)
     __attribute__ ((visibility ("hidden")));

/* Invoked by a timeout source when either an alarm should go off or
   when we need to look for additional upcoming events.  */
extern gboolean buzzer (gpointer data)
     __attribute__ ((visibility ("hidden")));

/* All EventSources are created with a toggle reference.  When the
   last user reference is dropped, this function is invoked and we
   grab a real reference and stick the event on the EDB->CACHE_LIST
   list.  If no user reference is taken before CACHE_EXPIRE seconds,
   the event is freed.  */
extern void event_source_toggle_ref_notify (gpointer data,
					    GObject *object,
					    gboolean is_last_ref)
     __attribute__ ((visibility ("hidden")));

#endif
