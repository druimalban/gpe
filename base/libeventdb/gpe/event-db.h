/*
 * Copyright (C) 2002 Philip Blundell <philb@gnu.org>
 * Copyright (C) 2006 Neal H. Walfield <neal@walfield.org>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version
 * 2 of the License, or (at your option) any later version.
 */

#ifndef EVENT_DB_H
#define EVENT_DB_H

#define SECONDS_IN_DAY (24*60*60)

#include <glib-object.h>
#include <glib.h>
#include <time.h>

#define MON  (1 << 0)
#define TUE  (1 << 1)
#define WED  (1 << 2)
#define THU  (1 << 3)
#define FRI  (1 << 4)
#define SAT  (1 << 5)
#define SUN  (1 << 6)

enum event_recurrence_type
{
  RECUR_NONE,
  RECUR_DAILY,
  RECUR_WEEKLY,
  RECUR_MONTHLY,
  RECUR_YEARLY
};

/**
 * *recur_t:
 *
 * Structure to hold recurrence information.
 */
typedef struct recur_s
{
  /* iCal's FREQ property.  */
  enum event_recurrence_type type;

  /* The number of times this recurrence set is expanded.  0 means
     there is no limit.  */
  unsigned int count;
  /* iCal's interval property: the number of units to skip.  If the
     recurrence type is RECUR_YEARLY then the first recurrence occurs
     INCREMENT years after the initial start.  */
  unsigned int increment;
  /* Only meaningful when TYPE is RECUR_WEEKLY: if bit 0 is set then
     occur on Mon, bit 1, Tue, etc.  */
  unsigned long daymask;

  /* A list of start times to exclude.  Must match the start of a
     recurrence exactly.  */
  GSList *exceptions;

  /* No recurrences beyond this time.  0 means forever.  */
  time_t end;
} *recur_t;

#define EVENT_TYPE (event_get_type ())
#define EVENT(obj) \
  (G_TYPE_CHECK_INSTANCE_CAST ((obj), EVENT_TYPE, struct _Event))
#define EVENT_CLASS(klass) \
  GTK_CHECK_CLASS_CAST (klass, EVENT_TYPE, struct _EventCLass)

struct _EventClass;
typedef struct _EventClass EventClass;

struct _Event;
typedef struct _Event Event;

extern GType event_get_type (void);

struct _EventDB;
typedef struct _EventDB EventDB;

#define TYPE_EVENT_DB             (event_db_get_type ())
#define EVENT_DB(obj) \
  (G_TYPE_CHECK_INSTANCE_CAST ((obj), TYPE_EVENT_DB, EventDB))
#define EVENT_DB_CLASS(klass) \
  (G_TYPE_CHECK_CLASS_CAST ((klass), TYPE_EVENT_DB, EventDBClass))
#define IS_EVENT_DB(obj) \
  (G_TYPE_CHECK_INSTANCE_TYPE ((obj), TYPE_EVENT_DB))
#define IS_EVENT_DB_CLASS(klass) \
  (G_TYPE_CHECK_CLASS_TYPE ((klass), TYPE_EVENT_DB))
#define EVENT_DB_GET_CLASS(obj) \
  (G_TYPE_INSTANCE_GET_CLASS ((obj), TYPE_EVENT_DB, EventDBClass))

/* Return GType of a view.  */
extern GType event_db_get_type (void);

/* An event db signals the "alarm-fired" signal when an event's alarm
   goes off (but only after a call to
   event_db_list_unacknowledged_alarms).  */
typedef void (*EventDBAlarmFiredFunc) (EventDB *, Event *);

/* Open a new database.  */
extern EventDB *event_db_new (const char *filename);

/* Search the event database EVD for the event with the uid UID.  */
extern Event *event_db_find_by_uid (EventDB *evd, guint uid);

/* Return the event whose alarm will first go off at or after NOW.  */
extern Event *event_db_next_alarm (EventDB *edb, time_t now);

/* Return the events in the event database EVD which occur between
   START and END.  */
extern GSList *event_db_list_for_period (EventDB *evd,
					 time_t start, time_t end);
/* Like event_db_list but returns the events whose alarm goes off
   between START and END.  */
extern GSList *event_db_list_alarms_for_period (EventDB *evd,
						time_t start, time_t end);
/* Like event_db_list_for_period but only for untimed events
   (i.e. which with a 0 length duration).  */
extern GSList *event_db_untimed_list_for_period (EventDB *evd,
						 time_t start, time_t end);

/* Returns the list of unacknowledged events since EDB was last such
   down and turns on the "alarm-fire" event.  Only call this function
   once at start up after having connected to the "alarm-fire"
   signal.  */
extern GSList *event_db_list_unacknowledged_alarms (EventDB *edb);

/* Create a new event in database EDB.  If EVENTID is NULL, one is
   fabricated.  */
extern Event *event_new (EventDB *edb, const char *event_id);

/* Flush event EV to the DB now.  */
extern gboolean event_flush (Event *ev);

/* Remove event EV from the underlying DB and dereference it.  */
extern gboolean event_remove (Event *ev);

/* Acknowledge that EV's alarm went off.  If EV does not have an alarm
   or it would go off in the future, does nothing.  */
extern void event_acknowledge (Event *ev);

/* g_object unref each Event * on the list and destroy the list.  */
extern void event_list_unref (GSList *l);

/* Return whether an event is a recurrent event.  */
extern gboolean event_is_recurrence (Event *ev) __attribute__ ((pure));
/* Return the event's recurrence data.  If this is changed, you must
   call event_flush.  */
extern recur_t event_get_recurrence (Event *ev) __attribute__ ((pure));
extern void event_clear_recurrence (Event *ev);

extern time_t event_get_start (Event *ev) __attribute__ ((pure));
extern void event_set_start (Event *ev, time_t time);

extern unsigned long event_get_duration (Event *ev) __attribute__ ((pure));
extern void event_set_duration (Event *ev, unsigned long duration);

extern unsigned long event_get_alarm (Event *ev) __attribute__ ((pure));
extern void event_set_alarm (Event *ev, unsigned long alarm);

/* Recurrence type.  */
extern enum event_recurrence_type event_get_recurrence_type (Event *ev)
     __attribute__ ((pure));
extern void event_set_recurrence_type (Event *ev,
				       enum event_recurrence_type type);

/* Start of the recurrence set.  */
extern time_t event_get_recurrence_start (Event *ev) __attribute__ ((pure));
extern void event_set_recurrence_start (Event *ev, time_t start);

/* End of the recurrence set.  */
extern time_t event_get_recurrence_end (Event *ev) __attribute__ ((pure));
extern void event_set_recurrence_end (Event *ev, time_t end);

extern gboolean event_get_untimed (Event *ev) __attribute__ ((pure));
extern void event_set_untimed (Event *ev, gboolean value);

extern unsigned long event_get_uid (Event *ev) __attribute__ ((pure));
extern const char *event_get_eventid (Event *ev) __attribute__ ((pure));


extern const char *event_get_summary (Event *ev) __attribute__ ((pure));
extern void event_set_summary (Event *ev, const char *summary);

extern const char *event_get_description (Event *ev) __attribute__ ((pure));
extern void event_set_description (Event *ev, const char *description);

extern const char *event_get_location (Event *ev) __attribute__ ((pure));
extern void event_set_location (Event *ev, const char *location);

extern const GSList *event_get_categories (Event *ev)
     __attribute__ ((pure));
extern void event_add_category (Event *ev, int category);
/* After calling this function, EV owns CATEGORIES.  If you need to
   continue to use CATEGORIES, pass a copy.  */
extern void event_set_categories (Event *ev, GSList *categories);

extern void event_add_exception (Event *ev, time_t start);

#endif
