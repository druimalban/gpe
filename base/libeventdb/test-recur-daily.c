/* test-recur-daily.c - Test daily recurrences.
   Copyright (C) 2006 Neal H. Walfield <neal@walfield.org>

   This file is part of GPE.

   GPE is free software; you can redistribute it and/or modify it
   under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   GPE is distributed in the hope that it will be useful, but WITHOUT
   ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
   or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
   License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111, USA. */

int do_test (int argc, char *argv[]);

#include "test-skeleton.h"

#include <stdio.h>
#include <error.h>
#include "gpe/event-db.h"

static int fail;

int
do_test (int argc, char *argv[])
{
  char *file;
  int fd;
  time_t now = time (NULL);

  /* Initialize the g_object system.  */
  g_type_init ();

  fd = create_temp_file (".", &file);
  /* event_db_new will open it itself.  */
  close (fd);

  EventDB *edb = event_db_new (file);
  if (! edb)
    error (1, 0, "evend_db_new");

  Event *ev = event_new (edb, NULL);
  event_set_start (ev, now);
#define D 100
  event_set_duration (ev, D);
  recur_t r = event_get_recurrence (ev);
  /* Every other day.  */
  r->type = RECUR_DAILY;
#define INCREMENT 2
  r->increment = INCREMENT;
#define COUNT 10
  r->count = COUNT;
  event_flush (ev);

  Event *ev2 = event_new (edb, NULL);
  event_set_start (ev2, now + D);
  event_set_duration (ev2, D);
  r = event_get_recurrence (ev2);
  /* Every day until 10 days from start.  */
  r->type = RECUR_DAILY;
#define END now + 10 * 24 * 60 * 60
  r->end = END;
  event_flush (ev2);

  int i;
  int expected_ev2_count;
  for (i = 0; i < COUNT * INCREMENT * 2; i ++)
    {
      /* XXX: Day light savings time...  */
      time_t t = now + i * 24 * 60 * 60;
      GSList *list = event_db_list_for_period (edb, now, t);

      GSList *e;
      int ev_count = 0;
      int ev_count2 = 0;
      for (e = list; e; e = g_slist_next (e))
	if (event_get_uid (e->data) == event_get_uid (ev))
	  ev_count ++;
	else if (event_get_uid (e->data) == event_get_uid (ev2))
	  ev_count2 ++;
	else
	  {
	    fail = 1;
	    printf ("Unknown and unexpected event with id %ld (i: %d)\n",
		    event_get_uid (e->data), i);
	  }

      if (i < COUNT * INCREMENT && ev_count != i / INCREMENT + 1)
	{
	  printf ("%d: Expected %d recurrences of ev but got %d (i: %d)\n",
		  __LINE__, i / INCREMENT + 1, ev_count, i);
	  fail = 1;
	}
      if (i >= COUNT * INCREMENT && ev_count != COUNT)
	{
	  printf ("%d: Expected %d recurrences of ev but got %d (i: %d)\n",
		  __LINE__, COUNT, ev_count, i);
	  fail = 1;
	}

      if (t <= END)
	expected_ev2_count = ev_count2;

      if (t <= END && ev_count2 != i)
	{
	  printf ("%d: Expected %d recurrences of ev2 but got %d (i: %d)\n",
		  __LINE__, i, ev_count2, i);
	  fail = 1;
	}
      if (t >= END && ev_count2 != expected_ev2_count)
	{
	  printf ("%d: Expected %d recurrences of ev2 but got %d (i: %d)\n",
		  __LINE__, expected_ev2_count, ev_count2, i);
	  fail = 1;
	}

      event_list_unref (list);
    }

  return fail;
}

