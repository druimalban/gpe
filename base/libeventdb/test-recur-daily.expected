2006-06-16 00:00:00 until 2006-06-17 23:59:59
2006-06-17 00:00:00 until 2006-06-18 23:59:59
2006-06-18 00:00:00 until 2006-06-19 23:59:59
2006-06-19 00:00:00 until 2006-06-20 23:59:59
2006-06-20 00:00:00 until 2006-06-21 23:59:59
2006-06-21 00:00:00: one
2006-06-21 23:50:00: two
2006-06-21 00:00:00 until 2006-06-22 23:59:59
2006-06-21 00:00:00: one
2006-06-21 23:50:00: two
2006-06-22 23:50:00: two
2006-06-22 00:00:00 until 2006-06-23 23:59:59
2006-06-22 23:50:00: two
2006-06-23 00:00:00: one
2006-06-23 23:50:00: two
2006-06-23 00:00:00 until 2006-06-24 23:59:59
2006-06-23 00:00:00: one
2006-06-23 23:50:00: two
2006-06-24 23:50:00: two
2006-06-24 00:00:00 until 2006-06-25 23:59:59
2006-06-24 23:50:00: two
2006-06-25 00:00:00: one
2006-06-25 23:50:00: two
2006-06-25 00:00:00 until 2006-06-26 23:59:59
2006-06-25 00:00:00: one
2006-06-25 23:50:00: two
2006-06-26 23:50:00: two
2006-06-26 00:00:00 until 2006-06-27 23:59:59
2006-06-26 23:50:00: two
2006-06-27 00:00:00: one
2006-06-27 23:50:00: two
2006-06-27 00:00:00 until 2006-06-28 23:59:59
2006-06-27 00:00:00: one
2006-06-27 23:50:00: two
2006-06-28 23:50:00: two
2006-06-28 00:00:00 until 2006-06-29 23:59:59
2006-06-28 23:50:00: two
2006-06-29 00:00:00: one
2006-06-29 23:50:00: two
2006-06-29 00:00:00 until 2006-06-30 23:59:59
2006-06-29 00:00:00: one
2006-06-29 23:50:00: two
2006-06-30 23:50:00: two
2006-06-30 00:00:00 until 2006-07-01 23:59:59
2006-06-30 23:50:00: two
2006-07-01 00:00:00: one
2006-07-01 00:00:00 until 2006-07-02 23:59:59
2006-07-01 00:00:00: one
2006-07-02 00:00:00 until 2006-07-03 23:59:59
2006-07-03 00:00:00: one
2006-07-03 00:00:00 until 2006-07-04 23:59:59
2006-07-03 00:00:00: one
2006-07-04 00:00:00 until 2006-07-05 23:59:59
2006-07-05 00:00:00: one
2006-07-05 00:00:00 until 2006-07-06 23:59:59
2006-07-05 00:00:00: one
2006-07-06 00:00:00 until 2006-07-07 23:59:59
2006-07-07 00:00:00: one
2006-07-07 00:00:00 until 2006-07-08 23:59:59
2006-07-07 00:00:00: one
2006-07-08 00:00:00 until 2006-07-09 23:59:59
2006-07-09 00:00:00: one
2006-07-09 00:00:00 until 2006-07-10 23:59:59
2006-07-09 00:00:00: one
2006-07-10 00:00:00 until 2006-07-11 23:59:59
2006-07-11 00:00:00 until 2006-07-12 23:59:59
2006-07-12 00:00:00 until 2006-07-13 23:59:59
2006-07-13 00:00:00 until 2006-07-14 23:59:59
2006-07-14 00:00:00 until 2006-07-15 23:59:59
2006-07-15 00:00:00 until 2006-07-16 23:59:59
